### ISSUE TITLE
New Version 12/07/2015 2.91

### ISSUE THREAD

**Issue by DavePlater** [delan/facepaste#27]
Facebook changed more of their code it seems.
I do not think this will fix for everyone, but it fixed mine. (en-US locale)

Since we are doing this in a patchwork kind of way, Firefox will complain about it NOT being signed.
You should be ok to use it anyway.
**Download link, drag and drop on the main folder, it should override any old facepaste versions:**
https://www.dropbox.com/s/510lgf7eyw7qczm/facepaste-2.91-fx.xpi?dl=0
*(Or use the 'Install Add-on From File' in the Add-ons Manager)*

Let me know how it works out for you all

**Comment by Xaemyl** [comment]
Thanks for taking this over! Your fork seems to be working so far with firefox 42 for both pages and profiles. Still doesnt download videos, however, but it gets the rest!

**Comment by TrueTenacity** [comment]
Thanks man,

New update works wonderfully, downloaded all of my gf's fb pictures :-)

**Comment by Unikat2000** [comment]
It doesn't work for me :(
"error: no photo found on photo page, are youaccepting third party cookies"
even when using an english version of firefox....you mean the system locale has to be  en-US?

**Comment by DavePlater** [comment]
My locale is en-US so the filenames and patterns from facebook would go along with that. I don't know if it is the issue, but facepaste looks for specific names of things to find files and if they don't match the pattern it fails.
I really haven't looked at that section of the code at all yet

**Comment by Xaemyl** [comment]
Yeah they must've made a code change, as I am now having the third party cookies error as well where it didnt before.

**Comment by interceptorchamp** [comment]
The new versions doesn´t download the pics, the following error message accours:
"(album 1 photo 1) successfully received photo page, creating photo file
(album 1 photo 1) error: no photo found on photo page, are youaccepting third party cookies?"

Firefox was already set to accept 3rd party cookies till they get invalid.
I set the options to acknowledge for each cookie, however it didn´t help even with manually accepting the cookies.

Do you have an idea what could help?
Can I manually set the firefox (German version) to en-US?

Any help will be much appreciated

Have a nice weekend

**Comment by DavePlater** [comment]
You guys who are seeing the "third party cookies" messages, could you open the BROWSER console windows (ctrl+shift+j i think) and run facepaste?
Let me know if it throws out any messages that say it came from facepaste (it should give a line # too)

**Comment by Xaemyl** [comment]
I get: "TypeError: img is null" from facepaste.

**Comment by DavePlater** [comment]
Did you get a line number?

**Comment by Xaemyl** [comment]
Was just about to edit my comment. lol here it is:  520: p.photourl = link ? link.href : img.src;

**Comment by DavePlater** [comment]
Yup the non-english facebook pages must have different names. That happens when it cannot find anything labeled "fbPhotoImage" or "fbPhotosPhotoActionsItem"

**Comment by Xaemyl** [comment]
The error occured on an english (en-US) profile. I'm thinking fb changed some code around or something, because it worked for a few hours after your original fork. Then again, my js is lacking. lol

**Comment by DavePlater** [comment]
Yeah I see it now. FB hadn't given me the new way photos are viewed until just today sometime. Working on it.

**Comment by Xaemyl** [comment]
Ahh looking over more of the code, I see what you're saying about the lables. heh

**Comment by DavePlater** [comment]
ok. i think i got both pictures and videos working. videos is still a little weird. Will update

**Comment by Xaemyl** [comment]
Excellent! :D

**Comment by interceptorchamp** [comment]
It works vor for pics, but it doesn´t work for videos.

It would be nice if the videos would be downloaded as well ... i know it worked about a year ago perfectly :/

Anyway many thanks for helping.

**Comment by DavePlater** [comment]
Yeah, i put that in my commit notes about extra videos, not sure where it gets the extra videos from. Facebook does the delay load until scroll thing which i think makes things funky.

If you're not getting videos, does it say it even finds them? it should say something like "found x videos" as well as "found y photos" now

**Comment by interceptorchamp** [comment]
Nope, always "found 0 videos", even in dedicated video folders.

Downloading the pics works fine, even in big albums with several 100 pics (no scoll issue).

**Comment by Unikat2000** [comment]
pics are working also for me again..thx  :)

**Comment by DavePlater** [comment]
I cannot find an album that has pics AND video to try and reproduce.
I you open a console window (CTRL+SHIFT+K I think), paste in this line of code:
```javascript
Array.prototype.slice.call(document.querySelectorAll("a")).filter( function(x) { return (x.rel=="theater")&&(x.href.indexOf("videos")!=-1);});
```
If you get "Array [ ]" then it didn't find any videos, check that.

Also try:
```javascript
Array.prototype.slice.call(document.querySelectorAll("a")).filter( function(x) { return (x.href.indexOf("videos")!=-1);});
```
To see if the videos show up in the Array []

**Comment by interceptorchamp** [comment]
Bote commands show the result
Array [  ]

I noticed that Firefox 43 de-installed facepaste as not-certified add-on. To run facepaste, the user needs to set "xpinstall.signatures.required" in about:config to "false", then it can be re-installed.
However this setting shall not be possible anymore in the upcoming Firefox 44 :/

**Comment by DavePlater** [comment]
The next thing would be to inspect element on the video link in the album.
Ctrl+shift+k to bring up the console, switch over to "Inspector" and click the button to the left (looks like a square with a mouse cursor in it)
Then click on a video link to see the source code for it. Try and copy/paste the whole <a> link to me.

For instance mine looks like:
```html
<a href="video_url" aria-label="Video" ajaxify="video_url" rel="theater">
	<span class="_5ig0 _47hq">
		<div class="uiScaledImageContainer _55sd" style="width:206px;height:206px;">
			<img class="scaledImageFitWidth img" src="img_src" style="top:-17px;" alt="" height="366" width="206"></div><div class="_3uzl _3uzq _47hr">
			<i class="_3uzm _3uzo"></i>
		</div>
		<div class="_5ig6 _50f3">0:32</div>
	</span>
</a>
```

**Comment by interceptorchamp** [comment]
<a href="/1471972006433081/videos/vb.1471972006433081/1472188756411406/?type=3&amp;video_source=pages_video_set" aria-label="Soaring at Schrattenbach" ajaxify="/1471972006433081/videos/vb.1471972006433081/1472188756411406/?type=3&amp;video_source=pages_video_set&amp;source=7" rel="theater">
<span class="_5ig0 _47hq"><div class="uiScaledImageContainer _55sd" style="width:206px;height:206px;">
<img class="scaledImageFitHeight img" src="https://scontent-frt3-1.xx.fbcdn.net/hvthumb-xta1/v/t15.0-10/p206x206/11331333_1472210979742517_1971295067_n.jpg?oh=7745fd8845fe713f4ae110c45fdf4f8a&amp;oe=56D96ECE" style="left:-80px;" alt="" height="206" width="366"></div><div class="_3uzl _3uzq _47hr"><i class="_3uzm _3uzo"></i></div><div class="_5ig6 _50f3">1:57
</div>
</span>
</a>

I hope it helps.

Thank you for your patience :)

**Comment by DavePlater** [comment]
Hmm it looks like I should drop the rel="theater" and just use href.indexOf("videos")!=-1
Both of ours also have type=3 in the href. I can probably work with that

**Comment by Xaemyl** [comment]
Looks like it's broken again. It throws the "file exists" error even if you're downloading into a new empty folder.

**Comment by DavePlater** [comment]
Yeah fb changed again, the name of the image was not being found on the webpage.
Which means it is trying to write a file with a blank name (which would just be the directory)
I am looking for a fix

**Comment by Xaemyl** [comment]
Thanks for all your work!

**Comment by Unikat2000** [comment]
thx for trying to help..... any progress so far?

**Comment by DavePlater** [comment]
The official mozilla addons has 2.91 which work if facebook has not updated you yet. On my github there is 2.91.1 which is still working for me as of yesterday. I have heard people have an even MORE different facebook that doesnt work still. If 2.91.1 doesnt work for you, have to wait for facebook to propagate the changes

**Comment by Xaemyl** [comment]
2.91.1 seems to be hanging on the "preparing" stage, so I'll wait a bit as maybe their changes havent propagated to me yet.

**Comment by interceptorchamp** [comment]
I had a look on your Github, but I can find only version 2.91 (without ".1").
Can you please point me to the most current version?

**Comment by Xaemyl** [comment]
https://github.com/DavePlater/facepaste/blob/master/facepaste.xpi (I believe lol)

**Comment by giticeman** [comment]
the new vr works with pic, but only for about ~50item .. and not all consecutives (it skip some item with the label "url: not yet know")
i've tried with different profiles, but all the time, it fetch correctly the various album, and pic or video into those, but stop download always after about 50 object.
FF 43

**Comment by Xaemyl** [comment]
Hmm. 2.91.1 for me (firefox 42) seems to hang (stuck on the grey "preparing" blocks) when trying to download photos, but downloads videos just fine. For both profiles and pages.

**Comment by Unikat2000** [comment]
I can confirm that "prepare block"  issue.....

**Comment by DavePlater** [comment]
The way I have to try and get photos from facebook is to "load the photo page" then wait and wait and wait until it fires all the javascript to make that "theater" screen popup (photo surrounded by black border). 
If it never loads a theater element, the page will never "give up". Do not have a work around for that.
Should mean that 2.91 will still work for you though.

**Comment by Xaemyl** [comment]
Ahh ok. Could mean I'm not letting it wait long enough. I'll let it wait and come back with my findings.

**Comment by Xaemyl** [comment]
Well, after letting it run for about 40 minutes, it never started to download (stuck on the preparing blocks), and 2.91 does the "file already exists" thing even though the folder is empty. Ahh well.

**Comment by Xaemyl** [comment]
Also upgraded to 43.0.1 and the same issue is happening. Curse my luck! hehe

**Comment by interceptorchamp** [comment]
Today I received the message for an update to Firefox 44.

As you are sure know, the tweak for running an unsigned add-on will not be available in Firefox 44.

Will we be able to run facepast after an update to Firefox 44, or should be still run Firefox 43.x.x for some time?

I run facepaste, but the error "file exists" still appears.
I can select the Folders which shall be downloaded, but that doesn´t help

Could you please help out?
Can we be of any support for you?

**Comment by captnfuture** [comment]
Same to me... Firefox 43.x and 44.0
Download to a new **empty** folder:

`Preparing to download 1 albums:
(album 1) Photos by noname
________________________________
(album 1) creating album folder
(album 1) fetching album index
(album 1) parsing album index
(album 1) found 62 photos
(album 1) found 2 videos
(album 1 photo 8) successfully received photo page, creating photo file
(album 1 photo 8) error creating photo file: **file exists**
(album 1 photo 7) successfully received photo page, creating photo file
(album 1 photo 7) error creating photo file: file exists
(album 1 photo 2) successfully received photo page, creating photo file
(album 1 photo 2) error creating photo file: file exists`

And so on...
But even no videos are selected, it downloads all videos, several times.
Can I help with some logs or something?
Thanks a lot for any help

**Comment by captnfuture** [comment]
Ah, something I have to say...
I can install only facepaste-2.91-fx.xpi from 2 month ago.
The recent one (facepaste.xpi) brings the error that the file is damaged.

**Comment by captnfuture** [comment]
Ok, now I downloaded the zip file and installed the recent facepaste.xpi from zip.
With that version I have the same problem like Xaemyl. It stucks on the grey "preparing" blocks after downloading few files.

**Comment by DavePlater** [comment]
Yeah, I do not have a fix for it. Yet. When I do, it will make it to the official addons in mozilla.
The problem is facebook's delayed loading thing. You load a picture page, some javascript fires and THEN the picture is shown. I do not have a good solution for this yet.

**Comment by LeaTex** [comment]
V2.91.1 is not working with FF 44.0.2.

It hangs out loading de album list. :/

**Comment by captnfuture** [comment]
Hi Dave
Perhaps I have a solution. Why not use the fb picture download functioality directly from album page instead of open the picturepage and search for the picture...

`function ILoadedAPage(myBrowser,p)
{
	regexParam = /\d{15}/; //this regex is not optimal but it works...
	var fbid = regexParam.exec(p.pageurl);
	var DownloadPath="https://www.facebook.com/photo/download/?fbid="+fbid;
	console.log("DownloadPath: "+DownloadPath);
	window.open(DownloadPath);// only for TESTING, proof of concept
`
Now, this method works seamlessly. The only problem is I dont now how to handle the download in background. Without the download window.

**Comment by DavePlater** [comment]
I ran into some issues trying to use that last time, such as could not get the name of the image correctly.
I will investigate again when I get some time.
I hope what you have works!

**Comment by lezantas26** [comment]
hello, i also have the same problem as captnfuture. is the developer still working on the issue, or could i get some help on how make it work? captnfuture seems like found a fix, but i dont know how to implement it myself 

please help :)

**Comment by captnfuture** [comment]
Hi Dave
I have another way to solve the problem and bypass the java scripts. Just go and download thru the mobile site: mbasic.facebook.com
There is no java at all.
Pick the username from url and switch to mobile site (https://mbasic.facebook.com/"username"/photos).

**Comment by DavePlater** [comment]
I do not see any download links using that url format. In fact I would say that is worse, as you do not even get the full size photos.

**Comment by Xaemyl** [comment]
And now that pages have changed, no more video downloads. :(
